doc/keynames/README.md

```diff
@@ -0,0 +1,477 @@
+If there are any discrepancies between this document, the Python implementation and the actual C implementation in [src/libs/elektra/keyname.c](../../src/libs/elektra/keyname.c), you should consider them as follows:
+
+1. The C implementation is optimized for speed and much harder to maintain.
+   It is mostly likely to be incorrect.
+2. In most cases, this document outranks the Python implementation.
+   There may, however, be cases where the language in this document was too vague and the Python implementation is actually correct.
+3. If two of the sources agree, the third one is probably incorrect.
+   Although again, if one of the agreeing sources is the C implementation it could still be the case that there is a mistake.
+
+In any case: If you find a discrepancy, please file a bug report at https://issues.libelektra.org/new.
+
+> _Note:_ Mistakes happen.
+> So there is no 100% always correct specification for Elektra's Key Names.
+> The goal is only to provide a reference that has a very high likelihood of being correct.
+
+_To Elektra Developers:_ Feel free to add any unclear or previous incorrect examples to the test cases in [tests/ctest/test_keyname.c](../../tests/ctest/test_keyname.c).
+These tests are very fast (1000+ test cases per second) and the more tests the better.
+
+## 1. Key Name Parts and Namespaces
+
+Before we get to Key Names proper, we need to talk about Key Name Parts and Namespaces.
+
+Each Key is part of one of these _Namespaces_:
+
+- Cascading
+- Meta
+- Spec
+- Proc
+- Dir
+- User
+- System
+- Default
+
+Each of these Namespaces has a very specific meaning, explained in [another section below](#12-namespaces-and-root-keys).
+
+Apart from the Namespace, a Key Name is just a series of zero or more _Key Name Parts_.
+Each Key Name Part is just an arbitrary (possibly empty) sequence of non-zero bytes.
+
+So without knowing anything about how Key Names are written, we could say that there is a Key in the Namespace "System" with the Key Name Parts "elektra", "version" and "info".
+
+> _Note:_ Not every such sequence, is a valid Key Name.
+> For more information see [section 4](#4-valid-and-invalid-key-names)
+
+### 1.1. Key Hierarchy
+
+You may have already seen elsewhere, that in Elektra Keys commonly look a bit like Unix paths:
+
+```
+/elektra/version/info
+```
+
+> _Note:_ How this representation works exactly and how namespaces come into play, will be explained in the next section.
+> For now, we only care that there is some similarity to Unix paths.
+
+This is no mistake.
+Elektra's _Key Database (KDB)_ is designed to resemble a Unix filesystem.
+In particular, the KDB has a similar hierarchy.
+More generally, all Key Names exhibit this hierarchy.
+By going back to thinking about a Key Name as a Namespace and a series of Key Name Parts, we can define this _Key Hierarchy_.
+
+Each Namespace has a separate hierarchy.
+The relation between these, will be explored in the next section.
```

Comment:
Overuse of forward references, give them all in the one note above (If you think it is needed. I think without any forward reference it is fine, too)